os/net/ipv6/multicast/mpl.c

```diff
@@ -0,0 +1,1807 @@
+  case 3:
+    /* 128 bit seed ID */
+    dst->s = 3;
+    for(i = 0; i < 16; i++) {
+      dst->id[i] = ptr[15 - i];
+    }
+    return;
+  default:
+    /* Invalid seed size */
+    return;
+  }
+}
+static void
+seed_id_host_to_net(void *dst, seed_id_t *src)
+{
+  /**
+   * Convert a seed id from our internal representation to network
+   * order and representation based on length s.
+   */
+  static uint8_t i;
+  static uint8_t *ptr;
+  ptr = dst;
+  switch(src->s) {
+  case 0:
+  case 3:
+    /* Both use 128 bit seed IDs and do exactly the same thing */
+    for(i = 0; i < 16; i++) {
+      /* Byte order must be swapped */
+      ptr[i] = src->id[15 - i];
+    }
+    return;
+  case 1:
+    /* 16 bit seed id */
+    ptr[0] = src->id[1];
+    ptr[1] = src->id[0];
+    return;
+  case 2:
+    /* 64 bit Seed ID */
+    for(i = 0; i < 8; i++) {
+      ptr[i] = src->id[7 - i];
+    }
+    return;
+  default:
+    /* Invalid seed size */
+    return;
+  }
+}
+static void
+update_seed_id(void)
+{
+  /* Load my seed ID into memory */
+#if MPL_SEED_ID_TYPE == 0
+  /* Copy seed ID from out link local ip address */
+  static uip_ds6_addr_t *my_ip6_addr;
+  my_ip6_addr = uip_ds6_get_global(ADDR_PREFERRED);
+  if(my_ip6_addr != NULL) {
+    seed_id_net_to_host(&local_seed_id, &my_ip6_addr->ipaddr, 0);
+  } else {
+    local_seed_id.s = MPL_SEED_ID_UNKNOWN;
+    LOG_DBG("Seed id not yet known.\n");
+    return;
+  }
+#elif MPL_SEED_ID_TYPE == 1
+  /* 16 bit seed id */
+  SEED_ID_S1(&local_seed_id, MPL_SEED_ID_L);
+#elif MPL_SEED_ID_TYPE == 2
+  /* 64 bit seed id */
+  SEED_ID_S2(&local_seed_id, MPL_SEED_ID_L);
+#elif MPL_SEED_ID_TYPE == 3
+  /* 128 bit seed id */
+  SEED_ID_S3(&local_seed_id, MPL_SEED_ID_L, MPL_SEED_ID_H);
+#endif
+
+  LOG_DBG("My seed id is ");
+  LOG_DBG_SEED(local_seed_id);
+  LOG_DBG_(" with S=%u\n", local_seed_id.s);
+}
+void
+icmp_out(struct mpl_domain *dom)
+{
+  uint8_t vector[32];
+  uint8_t vec_size;
+  uint8_t vec_len;
+  uint16_t payload_len;
+  uip_ds6_addr_t *addr;
+
+  LOG_INFO("MPL Control Message Out\n");
+
+  UIP_IP_BUF->vtc = 0x60;
+  UIP_IP_BUF->tcflow = 0;
+  UIP_IP_BUF->flow = 0;
+  UIP_IP_BUF->proto = UIP_PROTO_ICMP6;
+  UIP_IP_BUF->ttl = MPL_IP_HOP_LIMIT;
+
+  locsiptr = (struct seed_info *)UIP_ICMP_PAYLOAD;
+  payload_len = 0;
+
+  /* Set the source address to link local for now. If we need to, we can try changing it to global later */
+  uip_ip6addr_copy(&UIP_IP_BUF->destipaddr, &dom->ctrl_addr);
+  uip_ds6_select_src(&UIP_IP_BUF->srcipaddr, &UIP_IP_BUF->destipaddr);
+
+  /* Iterate over seed set to create payload */
+  for(locssptr = &seed_set[MPL_SEED_SET_SIZE - 1]; locssptr >= seed_set; locssptr--) {
+    if(SEED_SET_IS_USED(locssptr) && locssptr->domain == dom) {
+      locsiptr->min_seqno = locssptr->min_seqno;
+      SEED_INFO_CLR_LEN(locsiptr);
+      SEED_INFO_CLR_S(locsiptr);
+
+      /* Try setting our source address to global */
+      addr = uip_ds6_get_global(ADDR_PREFERRED);
+      if (addr) {
```

Comment:
code style: no space between ìf` and `(`. If seen other instanced of this in this PR, not only on `if` statements, please fix